src/protocols/secure_channel/CASESession.cpp

```diff
@@ -648,6 +687,83 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
     return CHIP_NO_ERROR;
 }
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+void CASESession::HandleConnectionAttemptComplete(Transport::ActiveTCPConnectionState * conn, CHIP_ERROR err)
+{
+    char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
+    VerifyOrReturn(conn != nullptr);
+    // conn->mAppState should not be NULL. SessionManager has already checked
+    // before calling this callback.
+    VerifyOrDie(conn->mAppState != nullptr);
+
+    // Transport::PeerAddress peerAddr = conn->mPeerAddr;
+    // peerAddr.ToString(peerAddrBuf);
+    conn->mPeerAddr.ToString(peerAddrBuf);
+
+    CASESession * caseSession = reinterpret_cast<CASESession *>(conn->mAppState->appContext);
+    VerifyOrReturn(caseSession != nullptr);
+
+    // Exit and disconnect if connection setup encountered an error.
+    SuccessOrExit(err);
+
+    ChipLogDetail(SecureChannel, "TCP Connection established with %s before session establishment", peerAddrBuf);
+
+    // Associate the connection with the current unauthenticated session for the
+    // CASE exchange.
+    caseSession->mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->SetTCPConnection(conn);
+
+    // Associate the connection with the current secure session that is being
+    // set up.
+    caseSession->mSecureSessionHolder.Get().Value()->AsSecureSession()->SetTCPConnection(conn);
+
+    // Send Sigma1 after connection is established for sessions over TCP
+    err = caseSession->SendSigma1();
+    SuccessOrExit(err);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(SecureChannel, "Connection establishment failed: %s", ErrorStr(err));
+
+        // Close the underlying connection
+        caseSession->mSessionManager->TCPDisconnect(conn);
+        caseSession->mPeerConnState = nullptr;
+
+        caseSession->Clear();
+    }
+}
+
+void CASESession::HandleConnectionClosed(Transport::ActiveTCPConnectionState * conn, CHIP_ERROR conErr)
```

Comment:
Will this get called if we ourselves call TCPDisconnect()?  This is very non-obvious to me so far.